rx_signal_processing/dsp.cu

```diff
@@ -133,17 +158,65 @@ namespace {
+      // each beam dir. Eigen overloads the * operator so we dont need to implement any matrix
+      // work ourselves.
+      auto result = phases * samps;  
       
-      auto beamformed_cast = reinterpret_cast<std::complex<float>*>(beamformed_samples.data());
+      // This piece of code just transforms the Eigen result back into our flat vector.
+      auto beamformed_cast = reinterpret_cast<std::complex<float>*>(result_ptr);
       Eigen::Map<Eigen::Matrix<std::complex<float>, Eigen::Dynamic, 
                                 Eigen::Dynamic, Eigen::RowMajor>>(beamformed_cast, result.rows(), 
                                                                   result.cols()) = result;
+    };
+
+    auto main_phase_offset = 0;
+    auto main_results_offset = 0;
+
+    // Now we calculate the offsets into the samples, phases, and results vector for each
+    // RX frequency. Each RX frequency could have a different number of beams, so we increment
+    // the phase and results offsets based off the accumulated number of beams. Once we have the
+    // offsets, we can call the beamforming lambda.
+    for (uint32_t rx_freq_num=0; rx_freq_num<beam_direction_counts.size(); rx_freq_num++) {
+
+      auto num_beams = beam_direction_counts[rx_freq_num];
+
+      // Increment to start of new frequency dataset.
+      auto main_sample_offset = num_samples * (num_main_ants + num_intf_ants) * rx_freq_num;
+      auto main_sample_ptr = filtered_samples.data() + main_sample_offset;
+
+      auto main_phase_ptr = phases.data() + main_phase_offset;
+
+      auto main_results_ptr = beamformed_samples_main.data() + main_results_offset;
+
+      beamform_from_offsets(main_sample_ptr, main_sample_ptr, main_results_ptr, 
+                            num_main_ants, num_beams);
+
+      // Only need to worry about beamforming the interferometer if its being used.
+      if (num_intf_ants > 0) {
+
+        // Skip the main array samples.
+        auto intf_sample_offset = main_sample_offset + (num_samples * num_main_ants);
+        auto intf_sample_ptr = filtered_samples.data() + intf_sample_offset;
+
+        auto intf_phase_offset = main_phase_offset + (num_beams * num_main_ants);
+        auto intf_phase_ptr = phases.data() + intf_phase_offset;
+
+        // Result offsets will be the same. Each main and intf will have one set of samples for
+        // each beam.
+        auto intf_results_offset = main_results_offset;
+        auto intf_results_ptr = beamformed_samples_intf.data() + intf_results_offset;
+
+        beamform_from_offsets(intf_sample_ptr, intf_phase_ptr, intf_results_ptr,
+                              num_intf_ants, num_beams);
+      }
+
       //Possibly non uniform striding means we incremement the offset as we go.
-      phase_offset += beam_direction_counts[beamdir_num] * num_antennas;
+      main_phase_offset += num_beams * (num_main_ants + num_intf_ants);
```

Comment:
This offset looks good though for using the data for the next frequency.

Reply:
okay.